### PR TITLE
catalog: add more detail to DescriptorNotFoundErrors

### PR DESCRIPTION
**Pull request by rafiss**
This should make it easier to debug some tests where this error is happening rarely

informs https://github.com/cockroachdb/cockroach/issues/104164

---

### catalog: use NewReferencedDescriptorNotFoundError

This is easier to remember than having to wrap the error at each call
site.

---

### catalog: add more detail to ErrDescriptorNotFound errors

Now the ID will always be shown when this error happens.

---

Release note: None

### CODE REVIEW

File: pkg/sql/catalog/errors.go [hunk 1]
```diff
@@ -89,10 +90,18 @@ func HasInactiveDescriptorError(err error) bool {
 // found with the given id.
 var ErrDescriptorNotFound = errors.New("descriptor not found")
```
Comment (chrisseto) [inline pkg/sql/catalog/errors.go:91]: Just a musing, but could we upgrade ErrDescriptorNotFound into a struct to discourage using it without providing additional details?

**Review by chrisseto — APPROVED** [review]
LGTM! Would it be worth it to add some tests to showcase the new details?

File: pkg/sql/catalog/errors.go (hunk 1 above)
Comment (rafiss) [reply pkg/sql/catalog/errors.go:91]: Yes we could, and that was my preferred first solution. The issue though is the ~50 or so usages of `errors.Is(err, catalog.ErrDescriptorNotFound)` throughout the codebase, which would no longer work if we changed it to be a struct. Making that change would make the backport slightly more risky.

**Comment by rafiss** [comment]
> Would it be worth it to add some tests to showcase the new details?

The existing tests seem to cover this (my first version of this PR hadn't yet updated them)

**Comment by rafiss** [comment]
tftr!

bors r+

**Pull request merged**